### PR TITLE
add the ability to decrypt TLS private keys

### PR DESCRIPTION
**Pull request by adshmh**
added the ability to accept a passphrase and decrypt a TLS private key, by adding a 'Passphrase' to the Options struct and the functionality required to decrypt TLS keys. 
This is needed to address https://github.com/docker/docker/issues/30935. A few rounds of reviews were done in https://github.com/docker/docker/pull/31364 on the code in this PR as part of stubbing out the overall implementation. 

Signed-off-by: Arash Deshmeh <adeshmeh@ca.ibm.com>

### CODE REVIEW

**Comment by adshmh** [comment]
@cyli, @tiborvass

File: tlsconfig/config.go [hunk 1]
```diff
@@ -127,6 +132,67 @@ func adjustMinVersion(options Options, config *tls.Config) error {
...
+		return nil, fmt.Errorf("no valid private key found")
+	}
+
+	var err error
+	if x509.IsEncryptedPEMBlock(pemBlock) {
+		keyBytes, err = x509.DecryptPEMBlock(pemBlock, []byte(passphrase))
+		if err != nil {
+			return nil, errors.Wrap(err, "private key is encrypted, but could not decrypt it")
+		}
+		keyBytes = pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: keyBytes})
```
Comment (cyli) [inline tlsconfig/config.go]: Usually this contains a type, such as `EC PRIVATE KEY` or `RSA PRIVATE KEY` - we can probably just keep the previous one:  `keyBytes = pem.EncodeToMemory(&pem.Block{Type: pemBlock.Type, Bytes: keyBytes})`

File: tlsconfig/config_test.go [hunk 2]
```diff
@@ -478,6 +486,42 @@ func TestConfigClientTLSValidClientCertAndKey(t *testing.T) {
...
+
+	tlsConfig, err := Client(Options{
+		CertFile:   cert,
+		KeyFile:    key,
+		Passphrase: "InvalidPassphrase",
+	})
+
+	if err == nil || tlsConfig != nil {
+		t.Fatal("Expected failure due to incorrect passphrase.")
+	}
```
Comment (cyli) [inline tlsconfig/config_test.go]: Do we want to assert that it is specifically a passphrase error?

```
if !IsErrEncryptedKey(err) || tlsConfig != nil ||  {
    ...
```

**Comment by cyli** [comment]
Thanks for working on this @adshmh!

File: tlsconfig/config_test.go (hunk 2 above)
Comment (adshmh) [inline tlsconfig/config_test.go]: Thank you for the review @cyli. Yes, this seems a better test. I have updated the PR to address both changes (pem.EncodeToMemory() and the specific passphrase error in the test.

**Pull request merged**